File: 02_Sorting_Searching/30_Binary_Search/BaseCode.java
public class BaseCode {
    
    public static int binarySearch(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;

        while (low <= high) {
            // (low + high) / 2를 하면 배열이 엄청 클 때 int 오버플로우가 발생할 수 있습니다.
            int mid = low + (high - low) / 2;

            if (arr[mid] == target) {
                return mid; // 찾았을 때 인덱스 반환
            } else if (arr[mid] > target) {
                high = mid - 1; // 타겟이 중간값보다 작으므로 왼쪽 탐색
            } else {
                low = mid + 1;  // 타겟이 중간값보다 크므로 오른쪽 탐색
            }
        }
        
        return -1; // 찾지 못함
    }

    public static void main(String[] args) {
        int[] data = {2, 4, 6, 8, 10, 12, 14, 16}; // 정렬된 배열
        int target = 10;
        
        int result = binarySearch(data, target);
        if (result != -1) {
            System.out.println("Target " + target + " is at index: " + result);
        } else {
            System.out.println("Target not found.");
        }
    }
}

File: 02_Sorting_Searching/30_Binary_Search/base_code.py
def binary_search(arr, target):
    """
    반복문을 이용한 이분 탐색 정석 구현 (배열은 반드시 정렬되어 있어야 함)
    """
    low = 0
    high = len(arr) - 1

    while low <= high:
        # (low + high) // 2 도 가능하지만, 
        # 다른 언어에서의 오버플로우를 방지하는 습관인 low + (high - low) // 2 도 좋습니다.
        mid = (low + high) // 2 

        if arr[mid] == target:
            return mid       # 타겟을 찾은 경우 인덱스 반환
        elif arr[mid] > target:
            high = mid - 1   # 타겟이 중간값보다 작으면 왼쪽 구간 탐색
        else:
            low = mid + 1    # 타겟이 중간값보다 크면 오른쪽 구간 탐색

    return -1 # 탐색 실패 시 -1 반환


# 사용 예시
if __name__ == "__main__":
    # 데이터는 정렬되어 있다고 가정
    data = [2, 4, 6, 8, 10, 12, 14, 16]
    target_val = 10
    
    result = binary_search(data, target_val)
    if result != -1:
        print(f"Target {target_val} is at index: {result}")
    else:
        print("Target not found.")